### ISSUE TITLE
Prevent triple-loading of collections when GenericFiles are reindexed

### ISSUE THREAD

**Issue by pbinkley** [ualbertalib/HydraNorth#969]
In the spike on Fedora API use #948 we found  that collections objects were being loaded three times during update_index - see jetty log:

```
INFO 14:52:47.788 (FedoraLdp) HEAD for: dev/j9/60/20/64/j9602064p
INFO 14:52:47.796 (FedoraLdp) GET resource 'dev/j9/60/20/64/j9602064p'
INFO 14:52:48.222 (FedoraLdp) HEAD for: dev/j9/60/20/64/j9602064p
INFO 14:52:48.231 (FedoraLdp) GET resource 'dev/j9/60/20/64/j9602064p'
INFO 14:52:48.684 (FedoraLdp) HEAD for: dev/j9/60/20/64/j9602064p
INFO 14:52:48.693 (FedoraLdp) GET resource 'dev/j9/60/20/64/j9602064p'
INFO 14:52:49.112 (FedoraLdp) HEAD for: dev/th/83/kz/38/th83kz38r/full_text
INFO 14:52:49.119 (FedoraLdp) HEAD for: dev/th/83/kz/38/th83kz38r/content
```

**Comment by mbarnett** [comment]
Peter, can you point me at the code that was triggering this?

I'm using:

```ruby
c = Collection.find('ff365526h')
c.update_index
```

with various communities and collections on my machine, and seeing this activity in the log:

```text
INFO 13:19:25.349 (FedoraLdp) HEAD for: dev/ff/36/55/26/ff365526h
INFO 13:19:25.365 (FedoraLdp) GET resource 'dev/ff/36/55/26/ff365526h'
INFO 13:19:26.994 (FedoraLdp) HEAD for: dev/e1/fe/d7/df/e1fed7df-ffb9-47a9-92da-033b33d4408d
INFO 13:19:27.007 (FedoraLdp) GET resource 'dev/e1/fe/d7/df/e1fed7df-ffb9-47a9-92da-033b33d4408d'
INFO 13:19:27.133 (FedoraLdp) HEAD for: dev/7e/5b/c1/e8/7e5bc1e8-cb09-41ca-b21b-abc3db6e6c2c
INFO 13:19:27.145 (FedoraLdp) GET resource 'dev/7e/5b/c1/e8/7e5bc1e8-cb09-41ca-b21b-abc3db6e6c2c'
INFO 13:19:27.273 (FedoraLdp) HEAD for: dev/92/af/e6/74/92afe674-99f0-4781-9fb4-8bffd12cbb5e
INFO 13:19:27.284 (FedoraLdp) GET resource 'dev/92/af/e6/74/92afe674-99f0-4781-9fb4-8bffd12cbb5e'
```

This seems legitimate. e1fed7df-ffb9-47a9-92da-033b33d4408d, 7e5bc1e8-cb09-41ca-b21b-abc3db6e6c2c, and 92afe674-99f0-4781-9fb4-8bffd12cbb5e appear to be serialized Hydra::AccessControls::Permission objects associated with ff365526h in this case.

I *did* once observe it double GETing a community when I visited its index page, but only once, and I'm not completely convinced I didn't hallucinate it, so if you can trigger it more reliably than I can, it might help.

One thing I am seeing a lot of is behaviour like this:

```text
INFO 10:37:00.887 (FedoraLdp) HEAD for: dev/f1/88/1k/89/f1881k89j
INFO 10:37:00.903 (FedoraLdp) GET resource 'dev/f1/88/1k/89/f1881k89j'
INFO 10:37:01.333 (FedoraLdp) HEAD for: dev/f1/88/1k/89/f1881k89j
```

The fact that it issues a second HEAD request after the GET makes me think that it's attempting to check if the resource was modified after the GET, to avoid holding a stale copy. Maybe the tripled HEAD-GET-HEAD-GET-HEAD-GET pattern you were seeing was related to your code somehow impacting the last-modified time ActiveFedora was seeing on the post-GET HEAD request? I need to dig further into the code to understand where that second HEAD comes from.

**Comment by pbinkley** [comment]
Matt,  I was tailing the jetty log while manipulating the object in the
rails console. There wasn't anything else going on on the vm at the time. I
was working on a GenericFile,  not a Collection as in your example. The
triple loading was therefore of the community/collection to which the
GenericFile belonged. Does that help?

**Comment by mbarnett** [comment]
Thanks, that might point me in the right direction

**Comment by pbinkley** [comment]
I'm hoping it will turn out that we don't need to load the collection at
all,  since the GF already knows the title and id of its collections to put
in the Solr doc, but there may be more going on.

**Comment by mbarnett** [comment]
Still looking into possible causes.

Things I've learned so far: 

- This behaviour consistently DOESN'T happen to newly uploaded objects with collections and communities.
- This behaviour consistently DOES happen to any object after changing EITHER its collection or community after uploading.
- This behaviour consistently DOESN'T happen to objects who have the same collection and community as they had when uploaded, even after editing other properties such as title.

For example, with a newly uploaded generic file xg94hp58h, the following code in the console:

```ruby
2.2.1 :018 > g = GenericFile.find('xg94hp58h')
 => #<GenericFile id: "xg94hp58h", mime_type: "image/jpeg", format_label: ["JPEG File Interchange Format"], file_size: ["489652"], last_modified: ["2016:02:19 11:37:30-07:00"], filename: ["koolaidmatt__2016-Feb-19.jpg"], original_checksum: ["97763125bb25c4a82beb7ae50a3ae7b7"], rights_basis: [], copyright_basis: [], copyright_note: [], well_formed: ["true"], valid: ["true"], status_message: [], file_title: [], file_author: [], page_count: [], file_language: [], word_count: [], character_count: [], paragraph_count: [], line_count: [], table_count: [], graphics_count: [], byte_order: ["big endian"], compression: ["JPEG (old-style)"], color_space: ["YCbCr"], profile_name: [], profile_version: [], orientation: [], color_map: [], image_producer: [], capture_device: [], scanning_software: [], exif_version: [], gps_timestamp: [], latitude: [], longitude: [], character_set: [], markup_basis: [], markup_language: [], bit_depth: [], channels: [], data_format: [], offset: [], frame_rate: [], label: "koolaidmatt__2016-Feb-19.jpg", depositor: "dittest@ualberta.ca", relative_path: "", import_url: nil, part_of: [], resource_type: ["Book Chapter"], title: ["asdf"], creator: ["asdfasdf"], contributor: [], description: [], rights: "", publisher: [], date_created: "", date_modified: Fri, 19 Feb 2016 18:37:24 +0000, subject: ["asdfasdf"], language: "French", identifier: [], related_url: "", source: "", proxy_depositor: nil, on_behalf_of: nil, date_uploaded: Fri, 19 Feb 2016 18:37:24 +0000, license: "http://creativecommons.org/licenses/by-nc-sa/4.0/", trid: nil, ser: nil, temporal: [], spatial: [], is_version_of: "", unicorn: nil, fedora3uuid: nil, fedora3handle: nil, ingestbatch: nil, hasCollection: ["Circumpolar Digital Image Collection"], belongsToCommunity: ["5t34sj58c"], hasCollectionId: ["vm40xr59d"], year_created: nil, remote_resource: nil, degree_grantor: nil, dissertant: nil, supervisor: [], committee_member: [], department: [], specialization: nil, date_submitted: Fri, 19 Feb 2016 18:37:24 +0000, date_accepted: nil, graduation_date: nil, alternative_title: [], thesis_name: nil, thesis_level: nil, proquest: nil, abstract: nil, batch_id: "0c483j46q", embargo_id: nil, lease_id: nil> 
2.2.1 :019 > g.update_index
 => {"responseHeader"=>{"status"=>0, "QTime"=>8}} 
```

produces the following FedoraLDP calls for the ```find```:

```text
INFO 12:15:02.821 (FedoraLdp) HEAD for: dev/xg/94/hp/58/xg94hp58h
INFO 12:15:02.831 (FedoraLdp) GET resource 'dev/xg/94/hp/58/xg94hp58h'
INFO 12:15:03.091 (FedoraLdp) HEAD for: dev/xg/94/hp/58/xg94hp58h/characterization
INFO 12:15:03.103 (FedoraLdp) GET resource 'dev/xg/94/hp/58/xg94hp58h/characterization'
```
and these FedoraLDP calls for the ```update_index```
```text
INFO 12:15:09.470 (FedoraLdp) HEAD for: dev/xg/94/hp/58/xg94hp58h/full_text
INFO 12:15:09.478 (FedoraLdp) HEAD for: dev/xg/94/hp/58/xg94hp58h/content
INFO 12:15:09.494 (FedoraLdp) HEAD for: dev/5e/a9/48/53/5ea94853-dae2-4c72-a51f-9fbaf12b8a10
INFO 12:15:09.505 (FedoraLdp) GET resource 'dev/5e/a9/48/53/5ea94853-dae2-4c72-a51f-9fbaf12b8a10'
```

Note that xg94hp58h has both a collection and community set, but that the update_index call, when xg94hp58h hasn't been altered at all, generates some checks for modification on the full_text and content objects, and re-GETs one permission object.

Contrast with the logs produced by the same sequence of calls for an item, h415p964f, that has had its Community changed after uploading:

```ruby
2.2.1 :023 >   g = GenericFile.find('h415p964f')
 => #<GenericFile id: "h415p964f", mime_type: "image/jpeg", format_label: ["JPEG", "JPEG image data"], file_size: ["269264"], last_modified: ["2016:02:19 11:57:00-07:00"], filename: ["ABE9EX1.jpg"], original_checksum: ["bc40e415317dd370b15c52f049c81c40"], rights_basis: [], copyright_basis: [], copyright_note: [], well_formed: ["true"], valid: ["false"], status_message: ["File does not begin with SPIFF, Exif or JFIF segment offset=73"], file_title: [], file_author: [], page_count: [], file_language: [], word_count: [], character_count: [], paragraph_count: [], line_count: [], table_count: [], graphics_count: [], byte_order: ["big endian"], compression: ["JPEG (old-style)", "Baseline"], color_space: [], profile_name: [], profile_version: [], orientation: [], color_map: [], image_producer: [], capture_device: [], scanning_software: [], exif_version: [], gps_timestamp: [], latitude: [], longitude: [], character_set: [], markup_basis: [], markup_language: [], bit_depth: [], channels: [], data_format: [], offset: [], frame_rate: [], label: "ABE9EX1.jpg", depositor: "dittest@ualberta.ca", relative_path: "", import_url: nil, part_of: [], resource_type: ["Report"], title: ["nnhjghjg"], creator: ["bijou"], contributor: [], description: [], rights: "", publisher: [], date_created: "", date_modified: Fri, 19 Feb 2016 12:01:09 +0000, subject: ["nkjkn"], language: "German", identifier: [], related_url: "", source: "", proxy_depositor: nil, on_behalf_of: nil, date_uploaded: Fri, 19 Feb 2016 18:56:55 +0000, license: "http://creativecommons.org/publicdomain/mark/1.0/", trid: nil, ser: nil, temporal: [], spatial: [], is_version_of: "", unicorn: nil, fedora3uuid: nil, fedora3handle: nil, ingestbatch: nil, hasCollection: ["Technical Reports"], belongsToCommunity: ["ff365526h"], hasCollectionId: ["3x816m604"], year_created: nil, remote_resource: nil, degree_grantor: nil, dissertant: nil, supervisor: [], committee_member: [], department: [], specialization: nil, date_submitted: Fri, 19 Feb 2016 18:56:55 +0000, date_accepted: nil, graduation_date: nil, alternative_title: [], thesis_name: nil, thesis_level: nil, proquest: nil, abstract: nil, batch_id: "1544bp163", embargo_id: nil, lease_id: nil> 
2.2.1 :024 > g.update_index
 => {"responseHeader"=>{"status"=>0, "QTime"=>8}}
```

Logs produced by the ```find```:

```text
INFO 12:19:57.862 (FedoraLdp) HEAD for: dev/h4/15/p9/64/h415p964f
INFO 12:19:57.874 (FedoraLdp) GET resource 'dev/h4/15/p9/64/h415p964f'
INFO 12:19:58.117 (FedoraLdp) HEAD for: dev/h4/15/p9/64/h415p964f/characterization
INFO 12:19:58.129 (FedoraLdp) GET resource 'dev/h4/15/p9/64/h415p964f/characterization'
INFO 12:19:58.168 (FedoraLdp) HEAD for: dev/3x/81/6m/60/3x816m604
INFO 12:19:58.179 (FedoraLdp) GET resource 'dev/3x/81/6m/60/3x816m604'
```

and for the ```update_index```:

```text
INFO 12:20:59.008 (FedoraLdp) HEAD for: dev/3x/81/6m/60/3x816m604
INFO 12:20:59.019 (FedoraLdp) GET resource 'dev/3x/81/6m/60/3x816m604'
INFO 12:20:59.235 (FedoraLdp) HEAD for: dev/3x/81/6m/60/3x816m604
INFO 12:20:59.262 (FedoraLdp) GET resource 'dev/3x/81/6m/60/3x816m604'
INFO 12:20:59.481 (FedoraLdp) HEAD for: dev/3x/81/6m/60/3x816m604
INFO 12:20:59.492 (FedoraLdp) GET resource 'dev/3x/81/6m/60/3x816m604'
INFO 12:20:59.739 (FedoraLdp) HEAD for: dev/h4/15/p9/64/h415p964f/full_text
INFO 12:20:59.748 (FedoraLdp) HEAD for: dev/h4/15/p9/64/h415p964f/content
INFO 12:20:59.765 (FedoraLdp) HEAD for: dev/e6/72/cf/6d/e672cf6d-9702-4421-a69b-ac0775d86a87
INFO 12:20:59.781 (FedoraLdp) GET resource 'dev/e6/72/cf/6d/e672cf6d-9702-4421-a69b-ac0775d86a87'
```

ActiveFedora is now producing a GET for the collection object on the ```find``` of the GenericFile, where it wasn't before, and is now performing the tripled GET of the collection object in the ```update_index``` call, where it made NO calls for the collection of an object previously.

This behaviour consistently happens to any object after changing EITHER its collection or community after uploading, and consistently DOESN'T happen to objects who have the same collection and community as they had when uploaded. Editing other properties such as title does not seem to produce this behaviour.

**Comment by pbinkley** [comment]
Just to be sure I understand: does changing an item's collection or community trigger the triple-loading on every ```update_index``` call for that object thereafter, or only for the next ```update_index``` after the change? My sense was that I was seeing the triple-load every time I reindexed an item, even without subsequent changes, but I may not have been watching carefully enough to notice.

**Comment by mbarnett** [comment]
On every ```update_index``` call thereafter. More specifically, ANY call to ```to_solr``` (which ```update_index``` uses), will exhibit that behaviour on every call once you've updated the collection or community.

**Comment by pbinkley** [comment]
So it sounds like maybe there's code intended to guarantee consistency between the Collection and the GenericFile that belongs to it, which triggers a check of the Collection when a change is made, but somehow fails to register the fact that the change has been checked. We had a problem in old ERA when Collections changed their names, and this change had to be propagated to their members (forcing a reindexing even though the member items hadn't changed). As I recall the solution has rather hacky. Maybe Sufia has something similar going on.

**Comment by mbarnett** [comment]
I thought it might be something like that, but I couldn't find any obvious inconsistencies between the "fresh" GenericFile and the "bugged" one. Tracing ```to_solr``` shows that the superfluous calls are actually coming from different methods inside of ActiveFedora (it GETs the collection twice while generating object_profile_ssm, and once again when generating rdf assertions), so the inconsistency in the bugged object is, I'm thinking, down in the RDF itself and somehow masked or resolved by the time the GenericFile representation has been constructed..

**Comment by mbarnett** [comment]
I'll need to spot-check some more objects to be sure, but this appears to be the crucial difference.

Good object:

```xml
<sv:property sv:name="ns002:hasCollectionId" sv:type="String" sv:multiple="true">
<sv:value>vm40xr59d</sv:value>
</sv:property>
<sv:property sv:name="ns002:belongsToCommunity" sv:type="String" sv:multiple="true">
<sv:value>5t34sj58c</sv:value>
</sv:property>
<sv:property sv:name="ns002:hasCollection" sv:type="String" sv:multiple="true">
<sv:value>Circumpolar Digital Image Collection</sv:value>
</sv:property>
```

Bugged object:

```xml
<sv:property sv:name="ns002:hasCollection_ref" sv:type="WeakReference" sv:multiple="true">
<sv:value>2b97e2ce-bd54-49bd-8afe-5ac4b3636c99</sv:value>
</sv:property>
<sv:property sv:name="ns002:hasCollectionId" sv:type="String" sv:multiple="true">
<sv:value>3x816m604</sv:value>
</sv:property>
<sv:property sv:name="ns002:belongsToCommunity" sv:type="String" sv:multiple="true">
<sv:value>ff365526h</sv:value>
</sv:property>
```

When we change the object's community or collection it seems to be replacing the hasCollection property with hasCollection_ref instead.

**Comment by pgwillia** [comment]
closed by PR #1002
